doc/coding_guide/Coding_guidelines.md

```diff
@@ -0,0 +1,650 @@
+
+-   Try to capture the balance between spreading headers to many source files, which makes dependency graph more complicated and accumulating many of the headers to some common include header, which may lead to significant code size inflation after preprocessing.
+
+-   Prefer to use forward declaration instead of including a header if you don't need complete type.
+
+-   Do not put using declarations in headers. It could lead to undesired scope expansion.
+
+**Macro**
+
+6.  Use macro only if you want some compile-time instrumentation.
+
+-   Other ways use functions, templates and constants.
+
+    Example:
+```
+    // Instead of
+    #define BLOCK_SIZE 1024
+
+    // Use
+    constexpr BLOCK_SIZE = 1024;
+```
+-   Remember when you pass something to macro, it will be used in the way you provided it and will be expanded textually.
+
+    Example:
+```
+    IPP_MAX(++i, j);
+    // will become
+    ((++i) > (j) ? (++i) : j) // twice incremented, it is an unwanted side effect
+```
+-   Use ```MFX_CHECK``` / ```MFX_RETURN``` / ```MFX_CHECK_STS``` macro for ```mfxStatus``` checks. It allows to turn on tracing.
+
+    Hint: Treat these macros like an ```assert``` -  expression have to be evaluated to true and no action is taken in that case, otherwise error is returned.
+
+    Do not put complex conditions into these macros it may reduce readability, if you need such conditions use the following approach:
+```
+    bool mem_pattern_valid = (par->IOPattern & MFX_IOPATTERN_OUT_VIDEO_MEMORY) && (par->IOPattern & MFX_IOPATTERN_OUT_SYSTEM_MEMORY));
+    MFX_CHECK(mem_pattern_valid, MFX_ERR_INVALID_VIDEO_PARAM);
+```
+    Do not put function calls inside these macro, it may lead to double call.
+
+    Example:
+```
+    MFX_CHECK_WITH_ASSERT(MFX_ERR_NONE == SetHRD(par, m_vaDisplay, m_vaContextEncode, m_hrdBufferId), MFX_ERR_DEVICE_FAILED);
+    // This code will call SetHRD twice in debug, because in debug expansion of MFX_CHECK_WITH_ASSERT has additional assert.
+    // It is definitely not what author of the code desired
+```
+-   Include guards must be used for every header.
+
+-   Include guards are preferable to ```#pragma once```, since the last one is not standardized. But all popular compilers supports it. Don't mix these two approaches, choose one.
+
+-   Stop (!!) inventing/using maximum/minimum macro. We already have ```IPP_MAX```, ```MFX_MAX```, ```UMC_MAX```, ```Max```. Use ```std::max```/```min``` instead.
+
+7.  Do not overuse conditional compilation.
+
+-   Do not leave dead code under ```#if 0```. Use comments to hide it from compiler (do not forget to verbose your intention in related comment).
+
+8.  Use macro when extend public API / functionality.
+
+-   Fix binary layout of newly defined structs with ```MFX_PACK_BEGIN_USUAL_STRUCT``` / ```MFX_PACK_BEGIN_STRUCT_W_PTR``` macro (they are expanded to proper ```#pragma pack``` directives) to guarantee compatibility with different compilers.
```

Comment:
Is that related to API support or library code support? If API support maybe it worths absolutely separate section in the document.